### ISSUE TITLE
Disabling Notifications sent via email while users are online

### ISSUE THREAD

**Issue by Semir1212** [humhub/humhub#3891]
### Is your feature request related to a problem? Please describe.

Email notifications are sent although users are logged in and actively using the software e.g. writing messages via the Mail module. In this case, an email would be prompted for every message received. 

### Describe the solution you'd like

Logged in users should not receive Notifications via email. 

### Describe alternatives you've considered

The problem with the solution could be that users could be logged in with other devices e.g. their mobile phone. The described way would be the easiest solution. However, a more complex solution would be preferable e.g.:

- Delayed notifications depending on whether the user has read the notification
- Delayed "summary notifications"
- Grouping of notifications
- Summaries of notifications indicating the importance of single notifications with a custom email layout

**Comment by fae7em** [comment]
We can create a job to check if there are any unseen notifications for each user in a period of for example every 30 minutes, and send all of them in an email (those that have email targets).

**Comment by luke-** [comment]
It is really not that easy to implement.

Would it be possible to simply skip the mail target for online users and create an new (Active)Job (e.g. NotificationMail) instead?
For this ActiveJob we could use the `delay` and `retries` to postpone the action until the user is offline.

https://github.com/yiisoft/yii2-queue/blob/master/src/Queue.php#L79
https://github.com/yiisoft/yii2-queue/blob/master/src/Queue.php#L130

**Comment by fae7em** [comment]
Exactly we should skip `mailTarget` in `NotificationManager` and just handle these kinds of notifications through a new job.
We can just skip `mailTarget`, and instead implement a new `NotificationMail` to check if there are any unseen notification with `mailTarget` for each user and send all of them (in a period of time) as a group. 
or we can push `mailTarget` notifications with a specific delay to the new `NotificationMail` job and there check if this notification is unseen, then send it. 

Can we realize if a user is online or not?
> Would it be possible to simply skip the mail target for online users and create a new (Active)Job (e.g. NotificationMail) instead?
> For this ActiveJob we could use the `delay` and `retries` to postpone the action until the user is offline.

**Comment by luke-** [comment]
I'm not sure that I fully understood the two options - would you like to do a small Draft PR?
The session table should be a good way to check if the user is online.

**Comment by fae7em** [comment]
No, actually those are similar to each other, except in the first option, we can send all unseen notifications as a group in one email for example every 30 minutes, and the second one sends them individually after the `delay` time and other conditions.

**Comment by buddh4** [comment]
Can't we somehow just delay mobile and mail notification at all and just sent them seperately in case they were not read already? I think this is how other platforms are handling (at least mobile notifications).

**Comment by luke-** [comment]
I would also prefer the solution with grouping. But we only have to check here how the layout in the e-mail could be and also if this is possible for all notification types.

**Comment by fae7em** [comment]
One of the approaches could be to send the email notifications as a group like `MailSummary` layout contain notification contents, and for mobile notification just send for example "You have 3 unseen notifications" or something like that.

**Comment by luke-** [comment]
I'm not sure about mobile notifications, but for e-mail it sounds good!

**Comment by buddh4** [comment]
I'd rather prefer only grouping notifications by means of the current notification grouping and not mixing up different kinds of notification. For example I like that my github Notifications in my mail client regarding a single issue are grouped together as one Thread.

**Comment by luke-** [comment]
Ok, there is a `group_key` which may can be used for this. (It's already used  in the Notification Overview.)

**Comment by fae7em** [comment]
We can send notifications with the same `group_key` as a group and send the other notifications individually.
Currently, post`s comments and likes have the same `group_key`, but mention has no `group_ky`.

**Comment by luke-** [comment]
Unfortunately, the `group_key` is not used that much - but we should add it where possible e.g. Mention is a good place!

**Comment by buddh4** [comment]
Shouldn't ther only be a single mention per post anyways?

**Comment by fae7em** [comment]
> Shouldn't there only be a single mention per post anyways?

You mean if a user is mentioned in a post more than once, he/she just get one notification for them?

**Comment by buddh4** [comment]
I mean I don't know how we should use the `group_key` for mentioning notifications. But thats another topic lets focus on this issue. Maybe we should play around with the queue delay: https://www.yiiframework.com/extension/yiisoft/yii2-queue/doc/api/2.0/yii-queue-queue#delay()-detail . But note:

> Important: Not all drivers support delayed running.

From https://www.yiiframework.com/extension/yiisoft/yii2-queue/doc/guide/2.2/en/usage

**Comment by fae7em** [comment]
> Important: Not all drivers support delayed running.

So what if a project uses a queue without delay support? Can we just send emails without delay in that cases (like the current situation)?
Or as I mentioned before, instead of using a queue delay, just create a cronjob to check the notification table and send the records that their created_at is between DATE1 and DATE2, and are un-seen.

**Comment by fae7em** [comment]
These drivers don't support delay:
`Synchronous`, `Gearman `, `Stomp` 

and these drivers support it:
`amqp_interop`, `beanstalk`, `db`, `file`, `redis`, `sqs`

**Comment by luke-** [comment]
For us the most important drivers are DB & Redis and for test cases InstantQueue. 
So there should be nothing against using the delay feature of the queue system.

**Comment by CarlSinclair** [comment]
Just wanted to suggest that in most cases, users prefer to be notified of each individual notification immediately. Isn't that the point of push notifications? For email, sure, they can be grouped and delayed. But not push. Yes the risk of getting spammed exists, but that's why settings and "turn off notifications for this post" exist.
